### PR TITLE
WIP: Hyperbola by foci passing through a point

### LINKED ISSUE #125: Function hyperbola(point F1, point F2, point M)?
I have found the ellipse function ellipse(point F1, point F2, point M). What is the similar function of hyperbola?

### PR DESCRIPTION
**Pull request by ivankokan**
Resolves #125.

### CODE REVIEW

**Comment by ivankokan** [comment]
I would call this a "blind commit", not tested. Please test and verify.

@quanhinh9 Can you test as well (at least copying the added function in your local environment)?

@pivaldi Considering the idea of having geometry module Index up-to-date, and possibly given here on github under Wiki (https://sourceforge.net/p/asymptote/discussion/409349/thread/f07c1fcf/), do you still have some scripts for generating the Index page I can start with?

**Comment by quanhinh9** [comment]
I use this function, it's ok.

`hyperbola hyperbola(point F1, point F2, point M)
{/*<asyxml></code><documentation>Return the hyperbola passing through 'M' whose the foci are 'F1' and 'F2'.</documentation></function></asyxml>*/
	point P[] = standardizecoordsys(false, F1, F2, M);
	real a = abs(abs(F1 - M) - abs(F2 - M));
	return hyperbola(F1, F2, finite(a) ? a/2 : a);
}`

**Comment by ivankokan** [comment]
I briefly tested it, looks good for non-degenerate cases. I will check for degenerate considering all conics, please stand by.

File: base/geometry.asy [hunk 1]
```diff
@@ -2548,7 +2548,7 @@ struct ellipse
   /*<asyxml><property type = "point" signature="F1,F2,C"><code></asyxml>*/
   restricted point F1,F2,C;/*<asyxml></code><documentation>Foci and center.</documentation></property><property type = "real" signature="a,b,c,e,p"><code></asyxml>*/
   restricted real a,b,c,e,p;/*<asyxml></code></property><property type = "real" signature="angle"><code></asyxml>*/
-  restricted real angle;/*<asyxml></code><documentation>Value is degrees(F1 - F2).</documentation></property><property type = "line" signature="D1,D2"><code></asyxml>*/
+  restricted real angle;/*<asyxml></code><documentation>Value is degrees(F2 - F1).</documentation></property><property type = "line" signature="D1,D2"><code></asyxml>*/
```
Comment (ivankokan) [inline base/geometry.asy]: now in correct order

File: base/geometry.asy [hunk 2]
```diff
@@ -2559,7 +2559,7 @@ struct ellipse
     this.F1 = P[0];
     this.F2 = P[1];
     this.C = (P[0] + P[1])/2;
-    this.angle = abs(P[1]-P[0]) < 10 * epsgeo ? 0 : degrees(P[1]-P[0]);
+    this.angle = degrees(F2 - F1, warn=false);
```
Comment (ivankokan) [inline base/geometry.asy]: a bit nicer way of handling congruent foci, i.e. circle

File: base/geometry.asy [hunk 3]
```diff
@@ -2606,7 +2606,7 @@ struct parabola
     this.a = distance(P[0], l)/2;
     this.p = 2 * a;
     this.V = 0.5 * (F + projection(D) * P[0]);
-    this.angle = degrees(F - V);
+    this.angle = degrees(F - V, warn=false);
```
Comment (ivankokan) [inline base/geometry.asy]: allow degenerate case when the focus lies on the directrix

File: base/geometry.asy [hunk 4]
```diff
@@ -2628,7 +2628,7 @@ struct hyperbola
     this.F1 = P[0];
     this.F2 = P[1];
     this.C = (P[0] + P[1])/2;
-    this.angle = degrees(F2 - F1);
+    this.angle = degrees(F2 - F1, warn=false);
```
Comment (ivankokan) [inline base/geometry.asy]: allow degenerate case when foci are congruent, i.e. hyperbola is a single point

**Comment by ivankokan** [comment]
Thanks @johncbowman. Merged content brings improvement and will not do any harm in non-degenerate cases. However, I suggest we reopen this PR and really handle degenerate cases in the future (that is why I put WIP prefix).

**Comment by quanhinh9** [comment]
I would like to suggest some new functions:
- tangent to conic and parallel with line.
- common tangent of two conics.
- some function arc for parabola, hyperbola.

**Comment by ivankokan** [comment]
@quanhinh9 Please submit new Issue for such enhancements.

**Pull request merged**